### ISSUE TITLE
"required" attribute of ChoiceTransformer (AbstractTransformer) not passed correctly to liform-react ChoiceWidget

### ISSUE THREAD

**Issue by kylemartin87** [Limenius/Liform#4]
Hi there

I've been attempting to implement Liform (through the Symfony Liform Bundle) together with the liform-react library. So far I've had good success and really appreciate your sharing this with the public.

I have, however, run into one issue:
The "required" attribute of the ChoiceTransformer (AbstractTransformer) is not passed correctly to the liform-react ChoiceWidget (or so I believe). Here's my best explanation of what is happening:

This is the json output of Liform:
```json
{  
   "announcementSchema":{  
      "title":"updaterbundle_announcement",
      "type":"object",
      "properties":{  
         "announcement":{  
            "type":"string",
            "title":"announcement",
            "attr":{  
               "id":"updater-textarea",
               "placeholder":"Enter line update ..."
            },
            "default":"Enter line update ...",
            "widget":"textarea",
            "propertyOrder":1
         },
         "announcementType":{  
            "enum":[  
               "One",
               "Two",
               "Three"
            ],
            "liform":{  
               "enum_titles":[  
                  "0",
                  "1",
                  "2"
               ]
            },
            "type":"string",
            "title":"Category",
            "attr":{  
               "placeholder":"-- Choose --"
            },
            "default":"-- Choose --",
            "propertyOrder":2
         }
      },
      "required":[  
         "announcement",
         "announcementType"
      ]
   },
   "announcementInitialValues":{  

   },
   "baseUrl":"\/updater\/",
   "location":"\/updater\/"
}
```
Everything looks correct according to Liform documentation. The two fields "announcement" and "announcementType" are contained in the "required" array.

Here is a snippet from renderField.js within liform-react that makes use of that Liform generated schema:
```javascript
var isRequired = exports.isRequired = function isRequired(schema, fieldName) {
    if (!schema.required) {
        return false;
    }
    return schema.required.indexOf(fieldName) != 1;
};
```

The problem is that the "schema" parameter passed in to that javascript function does not contain the "required" parameter since it is for an individual field (announcementType) and not the entire object. Here is what the "schema" parameter holds:
```json
attr: Object
    placeholder: "-- Choose --"
default: "-- Choose --"
enum: Array[3]
    0: "One"
    1: "Two"
    2: "Three"
    length: 3
liform: Object
    enum_titles: Array[3]
        0: "0"
        1: "1"
        2: "2"
        length: 3
propertyOrder: 2
title: "Category"
type: "string"
```

A possible fix would be to include an "addRequired" function to the Liform AbstractTransformer that looks something like this:
```php
    /**
     * @param mixed $form
     * @param array $schema
     *
     * @return array
     */
    protected function addRequired($form, $schema)
    {
        // If it has already been set then don't override it
        if (!isset($schema['required']) && $required = $form->getConfig()->getOption('required')) {
            $schema['required'] = [$form->getName()];
        }

        return $schema;
    }
```
And we'd add that into the addCommonSpecs function of course:
```php
    /**
     * @param mixed $form
     * @param array $schema
     * @param array $extensions
     * @param string $widget
     *
     * @return array
     */
    protected function addCommonSpecs($form, $schema, $extensions = [], $widget)
    {
        $schema = $this->addLabel($form, $schema);
        $schema = $this->addAttr($form, $schema);
        $schema = $this->addPattern($form, $schema);
        $schema = $this->addDefault($form, $schema);
        $schema = $this->addDescription($form, $schema);
        $schema = $this->addRequired($form, $schema);
        $schema = $this->addWidget($form, $schema, $widget);
        $schema = $this->applyExtensions($extensions, $form, $schema);

        return $schema;
    }
```

Does this make sense? Or is there a better way of solving this issue?

Kyle

**Comment by nacmartin** [comment]
Thanks for this detailed report @kylemartin87!

It is indeed a problem, yes.

I have moved the isRequired function to be called on `renderFields` instead of `renderField`(renderFields is called when the object is processed and it is rendering its children), so the boolean `required` is passed from the parent to their children.

What you suggest would also work, but the problem is that it is not how JSON Schema works, (Inded, my opinion is that JSON Schema should work as you describe, but I cannot change it :). You are supposed to provide an array of `required` fields in the parent object, not on its children.

(In fact I remember thinking about this and I even remember fixing this in the past, but I cannot find this fix in the git history). In any case, thanks for your report. This should be now fixed and tested for every case in liform-react.

I have released a new version of liform-readct with these changes https://github.com/Limenius/liform-react/releases/tag/v0.5.15
